engine/src/main/java/org/hibernate/validator/internal/cfg/context/TypeConstraintMappingContextImpl.java

```diff
@@ -118,24 +117,58 @@
 
 	@Override
 	public PropertyConstraintMappingContext property(String property, ElementType elementType) {
-		Contracts.assertNotNull( property, "The property name must not be null." );
 		Contracts.assertNotNull( elementType, "The element type must not be null." );
+
+		if ( !( ElementType.FIELD.equals( elementType ) || ElementType.METHOD.equals( elementType ) ) ) {
+			throw LOG.getElementTypeHasToBeFieldOrMethodException();
+		}
+
+		if ( ElementType.FIELD == elementType ) {
+			return field( property );
+		}
+		else {
+			return getter( property );
+		}
+	}
+
+	@Override
+	public PropertyConstraintMappingContext field(String property) {
+		Contracts.assertNotNull( property, "The property name must not be null." );
 		Contracts.assertNotEmpty( property, MESSAGES.propertyNameMustNotBeEmpty() );
```

Comment:
I think I would instead change `Contracts.assertNotEmpty()` to use `StringHelper.isNullOrEmptyString()`.